### PR TITLE
Add p2p communication

### PR DESCRIPTION
**Pull request by albestro**
This PR revamps a byproduct of @rasolca and "senderifise" it.

This is needed by distributed versions of BandToTridiagonal and its BackTransformation.

Additional changes:
- fix cmake-format problems on `master`
- fix build problem (missing header)

### CODE REVIEW

File: include/dlaf/communication/kernels/p2p.h
Comment (msimberg) [inline include/dlaf/communication/kernels/p2p.h:27]: I'm changing these in the other `communication/kernels` files at the moment and can do this later as well, but it could be updated here as well:
```suggestion
namespace dlaf::comm {
```

File: include/dlaf/communication/kernels/p2p.h
Comment (msimberg) [inline include/dlaf/communication/kernels/p2p.h:44]: Possibly...
```suggestion
void send(const matrix::Tile<const T, D>& tile, IndexT_MPI dest, IndexT_MPI tag,
          common::PromiseGuard<Communicator> pcomm, MPI_Request* req) {
  auto msg = comm::make_message(common::make_data(tile));
  DLAF_MPI_CHECK_ERROR(MPI_Isend(const_cast<T*>(msg.data()), msg.count(), msg.mpi_type(), dest, tag,
                                 pcomm.ref(), req));
}

DLAF_MAKE_CALLABLE_OBJECT(send);

// Non-blocking point to point receive
template <class T, Device D>
auto recv(const matrix::Tile<T, D>& tile, IndexT_MPI source, IndexT_MPI tag,
```
I know `sender` and `receiver` are already overloaded and the names would fit here as well, but considering that various MPI implementations call them `source` and `dest` those might be better names (https://www.mpich.org/static/docs/latest/www3/MPI_Isend.html).

**Review by msimberg — CHANGES_REQUESTED** [review]
Looks good to me in general, just a few minor requests.

File: test/unit/communication/test_comm_p2p.cpp
Comment (albestro) [inline test/unit/communication/test_comm_p2p.cpp:90]: For P2P communications it is not really needed to check that it works for both contiguous and strided memory, since that's a problem related to collective MPI operations.

In addition, it may be useful to test a more articulated use-case, e.g. using tagged communications inside the same communicator.

File: include/dlaf/communication/kernels/p2p.h
Comment (rasolca) [inline include/dlaf/communication/kernels/p2p.h:54]: How do you plan to use it if communications are not ordered?
with a shared_future like sender?

Comment (albestro) [reply include/dlaf/communication/kernels/p2p.h:54]: If you are talking about the problem that currently we implicitly require in our internal communication kernels to get a `PromiseGuard` instead of the actual `Communicator`, I agree with you and it is something I'd like to discuss with @msimberg in these days.

On the contrary, if it is not that, I'd ask you to explain a bit more in detail the problem you are thinking about.

Comment (albestro) [reply include/dlaf/communication/kernels/p2p.h:54]: @rasolca confirmed to me that the point was exactly that one. This is going to be addressed in #573.

File: test/unit/communication/test_comm_p2p.cpp
Comment (rasolca) [reply test/unit/communication/test_comm_p2p.cpp:90]: contiguous/strided is not related only to collectives...

Anyway testing mixing tags is a good idea.

Comment (albestro) [reply test/unit/communication/test_comm_p2p.cpp]: just added 😉 feel free to resolve this comment if you think it is enough.

**Review by rasolca — APPROVED** [review]

**Comment by rasolca** [comment]
bors try

File: include/dlaf/communication/kernels/p2p.h
Comment (msimberg) [inline include/dlaf/communication/kernels/p2p.h:66]: I'm happy to change this in #485 as well since I did similar changes there, but I'd try to have the `CommSender` as the first argument in `scheduleSend` and have the same order of arguments in `send` and `scheduleSend`.

Please feel free to leave this to me if your work is blocked on getting this PR in.

**Review by msimberg — APPROVED** [review]

**Pull request merged**
